extensions/markdown-language-features/src/languageFeatures/documentLinks.ts

```diff
@@ -130,300 +129,201 @@ export interface MdLinkDefinition {
 	const index = text.indexOf('#');
 	if (index < 0) {
 		return undefined;
 	}
-	return new vscode.Range(start.translate({ characterDelta: index + 1 }), end);
+	return new vscode.Range(linkRange.start.translate({ characterDelta: index + 1 }), linkRange.end);
 }
 
-function getLinkSourceFragmentInfo(document: SkinnyTextDocument, link: string, linkStart: vscode.Position, linkEnd: vscode.Position): { fragmentRange: vscode.Range | undefined; pathText: string } {
-	const fragmentRange = getFragmentRange(link, linkStart, linkEnd);
+function getLinkSourceFragmentInfo(document: SkinnyTextDocument, link: string, linkRange: vscode.Range): { fragmentRange: vscode.Range | undefined; pathText: string } {
+	const fragmentRange = getFragmentRange(link, linkRange);
 	return {
-		pathText: document.getText(new vscode.Range(linkStart, fragmentRange ? fragmentRange.start.translate(0, -1) : linkEnd)),
+		pathText: document.getText(new vscode.Range(linkRange.start, fragmentRange ? fragmentRange.start.translate(0, -1) : linkRange.end)),
 		fragmentRange,
 	};
 }
 
-const angleBracketLinkRe = /^<(.*)>$/;
+type TreeSitterLoader = () => Promise<Parser.Language>;
 
-/**
- * Used to strip brackets from the markdown link
- *
- * <http://example.com> will be transformed to http://example.com
-*/
-function stripAngleBrackets(link: string) {
-	return link.replace(angleBracketLinkRe, '$1');
+interface TreeSitterState {
+	readonly language: Parser.Language;
+	readonly linkQuery: Parser.Query;
 }
 
-const r = String.raw;
-
 /**
- * Matches `[text](link)` or `[text](<link>)`
+ * Stateless object that extracts link information from markdown files.
  */
-const linkPattern = new RegExp(
-	// text
-	r`(\[` + // open prefix match -->
-	/**/r`(?:` +
-	/*****/r`[^\[\]\\]|` + // Non-bracket chars, or...
-	/*****/r`\\.|` + // Escaped char, or...
-	/*****/r`\[[^\[\]]*\]` + // Matched bracket pair
-	/**/r`)*` +
-	r`\]` +
-
-	// Destination
-	r`\(\s*)` + // <-- close prefix match
-	/**/r`(` +
-	/*****/r`[^\s\(\)\<](?:[^\s\(\)]|\([^\s\(\)]*?\))*|` + // Link without whitespace, or...
-	/*****/r`<[^<>]*>` + // In angle brackets
-	/**/r`)` +
-
-	// Title
-	/**/r`\s*(?:"[^"]*"|'[^']*'|\([^\(\)]*\))?\s*` +
-	r`\)`,
-	'g');
-
-/**
-* Matches `[text][ref]` or `[shorthand]`
-*/
-const referenceLinkPattern = /(^|[^\]\\])(?:(?:(\[((?:\\\]|[^\]])+)\]\[\s*?)([^\s\]]*?)\]|\[\s*?([^\s\]]*?)\])(?![\:\(]))/gm;
+export class MdLinkComputer {
 
-/**
- * Matches `<http://example.com>`
- */
-const autoLinkPattern = /\<(\w+:[^\>\s]+)\>/g;
+	private readonly _treeSitter: Promise<TreeSitterState>;
 
-/**
- * Matches `[text]: link`
- */
-const definitionPattern = /^([\t ]*\[(?!\^)((?:\\\]|[^\]])+)\]:\s*)([^<]\S*|<[^>]+>)/gm;
+	constructor(
+		loadTreeSitter: TreeSitterLoader,
+		private readonly logger: ILogger,
+	) {
+		this._treeSitter = loadTreeSitter().then((lang): TreeSitterState => {
+			const linkQuery = lang.query(`([
```

Comment:
fyi - this will error with invalid queries. Maybe add some defence logic which might be helpful when grammars change